src/reader/decoder.rs

```diff
@@ -904,6 +894,46 @@ impl StreamingDecoder {
         Ok(())
     }
 
+    fn read_application_extension(&mut self) -> Option<Decoded> {
+        // the parser removes sub-block lenghts, so app name and data are concatenated
+        if let Some(&[first, second, ..]) = self.ext.data.strip_prefix(EXT_NAME_NETSCAPE) {
+            let repeat = u16::from(first) | u16::from(second) << 8;
+            return Some(Decoded::Repetitions(if repeat == 0 {
+                Repeat::Infinite
+            } else {
+                Repeat::Finite(repeat)
+            }));
+        } else if let Some(mut rest) = self.ext.data.strip_prefix(EXT_NAME_XMP) {
+            // XMP adds a "ramp" of 257 bytes to the end of the metadata to let the "pascal-strings"
+            // parser converge to the null byte. The ramp looks like "0x01, 0xff, .., 0x01, 0x00".
+            // For convenience and to allow consumers to not be bothered with this implementation detail,
+            // we cut the ramp.
+            const RAMP_SIZE: usize = 257;
+            if rest.len() >= RAMP_SIZE
+                && rest.ends_with(&[0x03, 0x02, 0x01, 0x00])
+                && rest[rest.len() - RAMP_SIZE..].starts_with(&[0x01, 0x0ff])
+            {
+                rest = &rest[..(rest.len() - RAMP_SIZE)];
+            }
+
+            self.xmp_metadata = Some(rest.to_vec());
+        } else if let Some(mut rest) = self.ext.data.strip_prefix(EXT_NAME_ICC) {
+            // The extension data always contains the length byte (see the XMP metadata extension
+            // for why that is necessary), so we need to remove it again for the ICC profile.
+            let mut icc_profile = Vec::new();
+            loop {
+                let block_length = *rest.first().unwrap_or(&0) as usize;
+                if block_length == 0 {
+                    break;
+                }
+                icc_profile.extend_from_slice(&rest[1..(1 + block_length)]);
+                rest.consume(1 + block_length);
```

Comment:
```suggestion
                let (subblock, tail) = rest.split_at(1 + block_length);
                icc_profile.extend_from_slice(&subblock[1..]);
                rest = tail;
```

Please don't do that with `std::io::BufRead`. With `1.87` we could also use [`split_off`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.split_off).

Reply:
Done.